File: src/0669_trim_a_binary_search_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} low
 * @param {number} high
 * @return {TreeNode}
 */
var trimBST = function (root, low, high) {
    if (!root) {
        return null;
    }

    if (root.val < low) {
        return trimBST(root.right, low, high);
    }

    if (root.val > high) {
        return trimBST(root.left, low, high);
    }

    root.left = trimBST(root.left, low, high);
    root.right = trimBST(root.right, low, high);

    return root;
};

module.exports = { trimBST };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(treeToString(trimBST(new TreeNode(1, new TreeNode(0, null, null), new TreeNode(2, null, null)), 1, 2)), treeToString(new TreeNode(1, null, new TreeNode(2, null, null))));
    assert.deepStrictEqual(treeToString(trimBST(new TreeNode(3, new TreeNode(0, null, new TreeNode(2, new TreeNode(1, null, null), null)), new TreeNode(4, null, null)), 1, 3)), treeToString(new TreeNode(3, new TreeNode(2, new TreeNode(1, null, null), null), null)));
}

File: src/0669_trim_a_binary_search_tree.rs
use std::cell::RefCell;
use std::rc::Rc;

struct Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        Self {
            val,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    pub fn trim_bst(
        root: Option<Rc<RefCell<TreeNode>>>,
        low: i32,
        high: i32,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        let Some(node) = root else {
            return None;
        };

        let val = node.borrow().val;
        if val > high {
            let left = node.borrow_mut().left.take();
            return Self::trim_bst(left, low, high);
        }

        if val < low {
            let right = node.borrow_mut().right.take();
            return Self::trim_bst(right, low, high);
        }

        let left = node.borrow_mut().left.take();
        node.borrow_mut().left = Self::trim_bst(left, low, high);
        let right = node.borrow_mut().right.take();
        node.borrow_mut().right = Self::trim_bst(right, low, high);
        Some(node)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cell::RefCell;
    use std::rc::Rc;

    fn node(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }

    fn tree_to_string(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        match root {
            Some(node) => {
                let node = node.borrow();
                format!(
                    "{}({},{})",
                    node.val,
                    tree_to_string(node.left.clone()),
                    tree_to_string(node.right.clone())
                )
            }
            None => "None".to_string(),
        }
    }

    #[test]
    fn main_asserts() {
        assert_eq!(
            tree_to_string(Solution::trim_bst(
                node(1, node(0, None, None), node(2, None, None)),
                1,
                2
            )),
            tree_to_string(node(1, None, node(2, None, None)))
        );
        assert_eq!(
            tree_to_string(Solution::trim_bst(
                node(
                    3,
                    node(0, None, node(2, node(1, None, None), None)),
                    node(4, None, None)
                ),
                1,
                3
            )),
            tree_to_string(node(3, node(2, node(1, None, None), None), None))
        );
    }
}
